### PR TITLE
Add possibilty to configure data source from file

### PR DESCRIPTION
**Pull request by golichenko02**

### CODE REVIEW

File: src/main/java/org/breskul/connectivity/datasource/DataSourcePropertyResolver.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+            if (isEmpty(url) || isEmpty(username) || isEmpty(password)) {
+                log.warn("Could not find one of the properties in application.properties file: {}, {}, {}. DataSource has not been initialized.", URL, USERNAME, PASSWORD);
+                return;
+            }
+            FileConfigurableDataSource.getInstance(url, username, password);
+        }
+    }
+
+    private void scanFile(String fileName) {
```
Comment (AndriiTsepukh) [inline src/main/java/org/breskul/connectivity/datasource/DataSourcePropertyResolver.java]: I would suggest reading properties file like (This is for correct species symbols handling and mirroring): 
Properties prop = new Properties();
ClassLoader loader = Thread.currentThread().getContextClassLoader();           
InputStream stream = loader.getResourceAsStream("/myProp.properties");
prop.load(stream);

File: src/main/java/org/breskul/connectivity/datasource/FileConfigurableDataSource.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        singletonInstance = new FileConfigurableDataSource(url, username, password);
+        return singletonInstance;
+    }
+
+    public static FileConfigurableDataSource getInstance() {
+        return singletonInstance;
+    }
+
+    private FileConfigurableDataSource(String url, String username, String password) {
```
Comment (AndriiTsepukh) [inline src/main/java/org/breskul/connectivity/datasource/FileConfigurableDataSource.java]: Why do we call it file-configurable data-source if we configure it by URL, username and password?

File: src/main/java/org/breskul/connectivity/datasource/DataSourcePropertyResolver.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    private Map<String, String> valueToPropertyName;
+
+    public DataSourcePropertyResolver() {
+        scanFile(DEFAULT_PROPERTIES_FILE_NAME);
+        initializeDataSource();
+    }
+
+    public DataSourcePropertyResolver(String fileName) {
+        scanFile(fileName);
+        initializeDataSource();
```
Comment (AndriiTsepukh) [inline src/main/java/org/breskul/connectivity/datasource/DataSourcePropertyResolver.java]: I would not initialize datasource inside the property resolver. For me, it's different responsibility. Property resolver can be more common class (for all properties).

Comment (golichenko02) [reply src/main/java/org/breskul/connectivity/datasource/DataSourcePropertyResolver.java]: - Extracted the connection pool and driver-agnostic logic to PooledDataSource
- Made PropertyResolver independent
- Added a bit documentation and logs

File: src/main/java/org/breskul/pool/PooledDataSource.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            throw new IllegalArgumentException(
+                    "Properties: %s, %s, %s must not be null. DataSource has not been initialized.".formatted(URL, USERNAME, PASSWORD));
+        }
+    }
+
+    private static PooledDataSource createDataSourceInstance(String url, String username, String password, int poolSize) {
+        return new PooledDataSource(url, username, password, poolSize);
+    }
+
+    private static PooledDataSource createDataSourceInstanceFromProperties() {
```
Comment (AndriiTsepukh) [inline src/main/java/org/breskul/pool/PooledDataSource.java:119]: I would create 1 more overloaded version of this method with the custom property file path.

**Review by AndriiTsepukh — APPROVED** [review]

File: src/main/java/org/breskul/pool/PooledDataSource.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    public String getDriverName() {
+        try {
+            return DriverManager.getDriver(this.url).toString();
+        } catch (SQLException e) {
+            throw new BoboException("Exception occurred when trying to retrieve driverName", e);
+        }
+    }
+
+    private PooledDataSource(String url, String username, String password) {
```
Comment (MishaBeheka) [inline src/main/java/org/breskul/pool/PooledDataSource.java:69]: Why did you create this constructor? It doesn't use anywhere.

File: src/test/resources/application.properties [hunk 6]
```diff
@@ -0,0 +1,4 @@
+url=jdbc:postgresql://localhost:5432/postgres
```
Comment (MishaBeheka) [inline src/test/resources/application.properties:1]: Why is this application.properties file in test resources?

**Pull request merged**